### PR TITLE
Add health endpoint

### PR DESCRIPTION
**Pull request by mpanik**
Basic health endpoint ~~with no details~~. I did remove `hello` route, since I found it obsolete (if the purpose was just checking responsiveness -- "health") 
 
Resolves: #91

### CODE REVIEW

File: lambdas/service/app.py [hunk 1]
```diff
@@ -27,11 +27,6 @@
 # how-to-write-docstring-for-url-parameters
 
 
-@app.route('/', cors=True)
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: One of the tests relies on this:

hca/azul/test/service/request_validation_test.py

File: lambdas/service/app.py [hunk 2]
```diff
@@ -54,6 +49,13 @@ def _get_pagination(current_request):
     return pagination
 
 
+@app.route('/health', methods=['GET'], cors=True)
+def health():
+    return {
+        'status': 'UP'
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Please add a check that ES is functional.

**Comment by mpanik** [comment]
@hannes-ucsc one think (probably issue with my deployment):

`/health` for `local indexer`:
```
{
    "status": "UP",
    "details": [
        {
            "elasticsearch": {
                "status": "UP",
                "details": {
                    "domain": "azul-index-jyo",
                    "endpoint": "search-azul-index-jyo-dbpfbbtisyd7ksom5elys4mksa.us-east-1.es.amazonaws.com"
                }
            }
        },
        {
            "queues": [
                {
                    "notifyQueue": {
                        "status": "UP",
                        "details": {
                            "url": "https://queue.amazonaws.com/861229788715/azul-notify-jyo",
                            "approximateNumberOfMessages": "0"
                        }
                    }
                },
                {
                    "failQueue": {
                        "status": "UP",
                        "details": {
                            "url": "https://queue.amazonaws.com/861229788715/azul-fail-jyo",
                            "approximateNumberOfMessages": "3"
                        }
                    }
                }
            ]
        }
    ]
}
```

`/health` for `deployed indexer` (https://indexer.jyo.dev.explore.data.humancellatlas.org/health):
```
{
    "status": "UP",
    "details": [
        {
            "elasticsearch": {
                "status": "UP",
                "details": {
                    "domain": "azul-index-jyo",
                    "endpoint": "search-azul-index-jyo-dbpfbbtisyd7ksom5elys4mksa.us-east-1.es.amazonaws.com"
                }
            }
        },
        {
            "queues": [
                {
                    "notifyQueue": {
                        "status": "UP",
                        "details": {
                            "url": "https://queue.amazonaws.com/861229788715/azul-notify-jyo",
                            "approximateNumberOfMessages": "0"
                        }
                    }
                },
                {
                    "failQueue": {
                        "status": "DOWN",
                        "details": {
                            "message": "The specified queue does not exist or you do not have access to it."
                        }
                    }
                }
            ]
        }
    ]
}
```

Any idea why deployed indexer have an issue with failQueue ?

**Comment by hannes-ucsc** [comment]
Look at how lambda-policy.json.template.py assigns permissions. Until now the indexer didn't need permissions to the fail queue because Amazon populates it.

File: src/azul/health.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            'elasticsearch': {
+                'status': 'DOWN',
+                'details': {
+                    'message': ex.response['Error']['Message'],
+                }
+            }
+        }
+
+
+def get_health_queues():
```
Comment (hannes-ucsc) [inline src/azul/health.py]: get_queue_health

File: src/azul/health.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            'notifyQueue': {
+                'status': 'DOWN',
+                'details': {
+                    'message': ex.response['Error']['Message'],
+                }
+            },
+        }
+
+
+def __get_health_fail_queue():
```
Comment (hannes-ucsc) [inline src/azul/health.py]: This function is 99% identical to the one above it.

https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

File: test/service/test_request_validation.py [hunk 5]
```diff
@@ -87,11 +87,11 @@ def tearDown(self):
         if self.server_thread.is_alive():
             self.fail('Thread is still alive after joining')
 
-    def test_hello_world(self):
+    def test_health(self):
         url = self.base_url
         response = requests.get(url)
         self.assertEqual(response.status_code, 200, response.json())
-        self.assertEqual({"Hello": "World!"}, response.json())
+        self.assertEqual({"Status": "UP"}, response.json())
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: I don't see how this test can pass because the response will have additional elements in it.

File: src/azul/health.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            'status': 'DOWN',
+            'message': ex.response['Error']['Message'],
+        }
+
+
+def get_queue_health():
+    notify_queue = __get_queue_health("notifyQueue", "azul-notify-{config.deployment_stage}".format(config=config))
+    fail_queue = __get_queue_health("failQueue", "azul-fail-{config.deployment_stage}".format(config=config))
+
+    return [
```
Comment (hannes-ucsc) [inline src/azul/health.py]: This is still a list of dicts. It should be just dictionary.

File: test/service/test_request_validation.py (hunk 5 above)
Comment (hannes-ucsc) [reply test/service/test_request_validation.py]: Does the test pass for you? Did you run it?

File: src/azul/health.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+
+def __get_queue_health(queue_name, sqs_queue_name):
+    try:
+        sqs = boto3.resource('sqs')
+        notify_queue = sqs.get_queue_by_name(QueueName=sqs_queue_name)
+
+        return {queue_name: {
+            'status': 'UP',
+            'approximateNumberOfMessages': notify_queue.attributes['ApproximateNumberOfMessages'],
```
Comment (hannes-ucsc) [inline src/azul/health.py]: I want us to use snake case wherever possible. Can you convert to this:

```
…
"messages" : {
    "queued": notify_queue.attributes['ApproximateNumberOfMessages'],
    "delayed": notify_queue.attributes['ApproximateNumberOfMessagesDelayed'],
    "invisible": notify_queue.attributes['ApproximateNumberOfMessagesNotVisible'],
}
…
```

File: src/azul/health.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+from botocore.exceptions import ClientError
+
+from azul import config
+
+
+def get_elasticsearch_health():
+    try:
+        es = boto3.client('es')
+        elasticsearch = es.describe_elasticsearch_domain(DomainName=config.es_domain)
+
```
Comment (hannes-ucsc) [inline src/azul/health.py]: I'd like to see an actual ping() using the ES client.

http://elasticsearch-py.readthedocs.io/en/master/api.html#elasticsearch.Elasticsearch.ping

File: test/service/test_request_validation.py (hunk 5 above)
Comment (mpanik) [reply test/service/test_request_validation.py]: @hannes-ucsc I forgot to update it. Anyway I'm not able to run tests. I've tried running them directly in PyCharm and also `make test`. Both options fails before getting to testing. If none of the options I've mentioned is default, we should have it covered in readme.

```
Testing started at 8:02 PM ...
/home/milan/.virtualenvs/.azul/bin/python /home/milan/.local/share/JetBrains/Toolbox/apps/PyCharm-P/ch-0/182.3684.100/helpers/pycharm/_jb_pytest_runner.py --target test_request_validation.py::FacetNameValidationTest.test_health
Launching pytest with arguments test_request_validation.py::FacetNameValidationTest::test_health in /home/milan/Projects/human-cell-atlas/azul/test/service

============================= test session starts ==============================
platform linux -- Python 3.7.0, pytest-3.2.3, py-1.5.4, pluggy-0.4.0
rootdir: /home/milan/Projects/human-cell-atlas/azul/test/service, inifile:
collected 1 item
test_request_validation.py EERROR:shared:Docker container took more than a minute to set up

test setup failed
cls = <class 'service.test_request_validation.FacetNameValidationTest'>

    @classmethod
    def setUpClass(cls):
>       super().setUpClass()

test_request_validation.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
__init__.py:17: in setUpClass
    super().setUpClass()
../shared.py:39: in setUpClass
    cls._wait_for_es()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'service.test_request_validation.FacetNameValidationTest'>

    @classmethod
    def _wait_for_es(cls):
        patched_log_level = logging.WARNING if logger.getEffectiveLevel() <= logging.DEBUG else logging.ERROR
        start_time = time.time()
        with patch.object(logging.getLogger('elasticsearch'), 'level', new=patched_log_level):
            while not cls.es_client.ping():
                if time.time() - start_time > 60:
                    logger.error('Docker container took more than a minute to set up')
>                   raise AssertionError
E                   AssertionError

../shared.py:49: AssertionError


==================================== ERRORS ====================================
____________ ERROR at setup of FacetNameValidationTest.test_health _____________

cls = <class 'service.test_request_validation.FacetNameValidationTest'>

    @classmethod
    def setUpClass(cls):
>       super().setUpClass()

test_request_validation.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
__init__.py:17: in setUpClass
    super().setUpClass()
../shared.py:39: in setUpClass
    cls._wait_for_es()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'service.test_request_validation.FacetNameValidationTest'>

    @classmethod
    def _wait_for_es(cls):
        patched_log_level = logging.WARNING if logger.getEffectiveLevel() <= logging.DEBUG else logging.ERROR
        start_time = time.time()
        with patch.object(logging.getLogger('elasticsearch'), 'level', new=patched_log_level):
            while not cls.es_client.ping():
                if time.time() - start_time > 60:
                    logger.error('Docker container took more than a minute to set up')
>                   raise AssertionError
E                   AssertionError

../shared.py:49: AssertionError
---------------------------- Captured stderr setup -----------------------------
ERROR:shared:Docker container took more than a minute to set up
========================== 1 error in 174.52 seconds ===========================
Process finished with exit code 0
```

**Comment by hannes-ucsc** [comment]
Oh, and add a test, please. To FacetNameValidationTest for the time being.

File: test/service/test_request_validation.py (hunk 5 above)
Comment (hannes-ucsc) [reply test/service/test_request_validation.py]: When you run `docker ps` while the test is spinning up, do you see a container for elasticsearch? If so, try attaching to it and see if it prints anything interesting. Also `docker logs` might give some insight.

Comment (hannes-ucsc) [reply test/service/test_request_validation.py]: @GPelayo is currently working on making `make test` functional again. The currently prescribed method for running tests is PyCharm with envhook installed in the virtualenv. @achave11 is adding docs about that to the README.

I do remember us installing envhook into your virtualenv while we were troubleshooting the project setup on your system but I don't see its effect in your log above. Without envhook installed, tests won't work either but they would fail at a later stage.

@alex-hancock does the error above look familiar?

Comment (alex-hancock) [reply test/service/test_request_validation.py]: Not at all. @mpanik do you have Docker installed on your machine? If you haven't installed it from [here](https://www.docker.com/community-edition), you might do that and try it again. I don't know why else it would take so long to set up.

File: src/azul/health.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        'status': 'UP' if is_es_reachable else 'DOWN',
+        'domain': config.es_domain
+    }
+    if not is_es_reachable:
+        result['message'] = 'Unable to reach the host'
+    return result
+
+
+def get_queue_health():
+    notify_queue_name = "notifyQueue"
```
Comment (hannes-ucsc) [inline src/azul/health.py]: I mentioned this before in the other PR: Generally, please don't extract expressions that are only used once. With a dict comprehension this can be made into a one liner.

File: src/azul/health.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    return {
+        notify_queue_name: __get_queue_health("azul-notify-{config.deployment_stage}".format(config=config)),
+        fail_queue_name: __get_queue_health("azul-fail-{config.deployment_stage}".format(config=config))
+    }
+
+
+def __get_queue_health(queue_name):
+    try:
+        sqs = boto3.resource('sqs')
```
Comment (hannes-ucsc) [inline src/azul/health.py]: Making this a nested function would allow you to share the SQS resource between invocations and eliminate the need for the `__` "protected" convention. See my patch.

File: test/service/test_request_validation.py [hunk 11]
```diff
@@ -87,11 +88,28 @@ def tearDown(self):
...
+
+    def test_health_es_unreachable(self):
+        self.env = EnvironmentVarGuard()
+        self.env.set('AZUL_ES_ENDPOINT', 'nonexisting-index.us-east-1.es.amazonaws.com:80')
+
+        url = self.base_url + "health"
+        response = requests.get(url)
+        expected_response = {
+            'status': 'UP',
+            'elasticsearch': {'domain': AzulConfig().es_domain, 'message': 'Unable to reach the host', 'status': 'DOWN'}
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Please wrap the nested dict literal.

File: test/service/test_request_validation.py [hunk 12]
```diff
@@ -87,11 +88,28 @@ def tearDown(self):
...
+        }
+        self.assertEqual(response.status_code, 200, response.json())
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: This is bad. Please see my comment in the other PR.

Use `response.raise_for_status()`

File: test/service/test_request_validation.py [hunk 13]
```diff
@@ -87,11 +88,28 @@ def tearDown(self):
...
-    def test_hello_world(self):
-        url = self.base_url
+    def test_health_es_reachable(self):
+        url = self.base_url + "health"
         response = requests.get(url)
+        expected_response = {
+            'status': 'UP',
+            'elasticsearch': {'domain': AzulConfig().es_domain, 'status': 'UP'}
+        }
         self.assertEqual(response.status_code, 200, response.json())
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: This is bad. Please see my comment in the other PR.

Use `response.raise_for_status()`

File: src/azul/__init__.py [hunk 14]
```diff
@@ -27,6 +30,27 @@ def es_endpoint(self) -> Tuple[str, int]:
             host, _, port = es_endpoint.partition(':')
             return host, int(port)
 
+    def es_connection(self):
```
Comment (hannes-ucsc) [inline src/azul/__init__.py]: I think this is too much for Config. See #184.

Comment (hannes-ucsc) [reply src/azul/__init__.py]: Rename to `es_client()`

File: src/azul/__init__.py [hunk 15]
```diff
@@ -27,6 +30,27 @@ def es_endpoint(self) -> Tuple[str, int]:
...
+        host = config.es_endpoint[0]
+        port = config.es_endpoint[1]
+        ssl_required = True if port == 443 else False
+        if ssl_required:
+            aws_auth = AWSRequestsAuth(
```
Comment (hannes-ucsc) [inline src/azul/__init__.py]:
```
Aligned indent everywhere.

aws_auth = AWSRequestsAuth(aws_host=host,
                           aws_region='us-east-1',
                           aws_service='es',
                           **boto_utils.get_credentials())
```

File: src/azul/health.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        result['message'] = 'Unable to reach the host'
+    return result
+
+
+def get_queue_health():
+    notify_queue_name = "notifyQueue"
+    fail_queue_name = "failQueue"
+
+    return {
+        notify_queue_name: __get_queue_health("azul-notify-{config.deployment_stage}".format(config=config)),
```
Comment (hannes-ucsc) [inline src/azul/health.py]: f-strings instead of `.format()` please.

Comment (hannes-ucsc) [reply src/azul/health.py]: Please use this

```diff
Index: src/azul/health.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/azul/health.py	(revision 92e61fcd56bbae28f456f2d72117abff5e998786)
+++ src/azul/health.py	(date 1533934040000)
@@ -18,30 +18,23 @@
 
 
 def get_queue_health():
-    notify_queue_name = "notifyQueue"
-    fail_queue_name = "failQueue"
-
-    return {
-        notify_queue_name: __get_queue_health("azul-notify-{config.deployment_stage}".format(config=config)),
-        fail_queue_name: __get_queue_health("azul-fail-{config.deployment_stage}".format(config=config))
-    }
+    sqs = boto3.resource('sqs')
 
-
-def __get_queue_health(queue_name):
-    try:
-        sqs = boto3.resource('sqs')
-        queue = sqs.get_queue_by_name(QueueName=queue_name)
-
-        return {
-            'status': 'UP',
-            'messages': {
-                'queued': queue.attributes['ApproximateNumberOfMessages'],
-                'delayed': queue.attributes['ApproximateNumberOfMessagesDelayed'],
-                'invisible': queue.attributes['ApproximateNumberOfMessagesNotVisible']
-            }
-        }
-    except ClientError as ex:
-        return {
-            'status': 'DOWN',
-            'message': ex.response['Error']['Message']
-        }
+    def queue_health(queue_name):
+        try:
+            queue = sqs.get_queue_by_name(QueueName=f"azul-{queue_name}-{config.deployment_stage}")
+            return {
+                'status': 'UP',
+                'messages': {
+                    'queued': queue.attributes['ApproximateNumberOfMessages'],
+                    'delayed': queue.attributes['ApproximateNumberOfMessagesDelayed'],
+                    'invisible': queue.attributes['ApproximateNumberOfMessagesNotVisible']
+                }
+            }
+        except ClientError as ex:
+            return {
+                'status': 'DOWN',
+                'message': ex.response['Error']['Message']
+            }
+
+    return {queue_name: queue_health(queue_name) for queue_name in 'notify', 'fail'}
```

File: test/service/test_request_validation.py [hunk 17]
```diff
@@ -87,11 +88,28 @@ def tearDown(self):
...
-        self.assertEqual({"Hello": "World!"}, response.json())
+        self.assertEqual(response.json(), expected_response)
+
+    def test_health_es_unreachable(self):
+        self.env = EnvironmentVarGuard()
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Use unittest.mock.patch.dict(). See the other PR for explanation and example.

**Comment by hannes-ucsc** [comment]
And resolve conflicts please.

**Comment by mpanik** [comment]
@hannes-ucsc:
**service**:
```
{
    "status": "UP",
    "elasticsearch": {
        "status": "UP",
        "domain": "azul-index-jyo"
    }
}
```

**indexer**:
```
{
    "status": "UP",
    "elasticsearch": {
        "status": "UP",
        "domain": "azul-index-jyo"
    },
    "queues": {
        "notify": {
            "status": "UP",
            "messages": {
                "queued": "0",
                "delayed": "0",
                "invisible": "0"
            }
        },
        "fail": {
            "status": "UP",
            "messages": {
                "queued": "2",
                "delayed": "0",
                "invisible": "0"
            }
        }
    }
}
```

File: src/azul/es.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
+from elasticsearch import Elasticsearch, RequestsHttpConnection
+from aws_requests_auth import boto_utils
+from aws_requests_auth.aws_auth import AWSRequestsAuth
+
+from azul import config
+
+
+def es_client():
+    host = config.es_endpoint[0]
```
Comment (hannes-ucsc) [inline src/azul/es.py]: This could have been a tuple destructuring assignment:

`host, port = config.es_endpoint`

File: test/service/test_request_validation.py [hunk 19]
```diff
@@ -49,14 +72,14 @@ def test_bad_single_filter_facet_of_specimen(self):
 
     def test_bad_multiple_filter_facet_of_specimen(self):
         url = self.base_url + "repository/specimens?from=1&size=1" \
-                               "&filters={'file':{'bad-facet':{'is':['fake-val']},'bad-facet2':{'is':['fake-val2']}}}"
```
Comment (hannes-ucsc) [inline test/service/test_request_validation.py]: Ideally these changes should be a separate commit.

**Comment by hannes-ucsc** [comment]
I also squashed the two commits into one.
